### ISSUE TITLE
Exceptions not thrown when multiplying scalar values of different batch sizes

### ISSUE THREAD

**Issue by huzecong** [clab/dynet#1004]
Consider the following example:

```python
import dynet as dy

x = dy.inputTensor([1, 2, 3], batched=True)
print(x.dim())  # ((1,), 3)

y = dy.inputTensor([1, 2], batched=True)
print(y.dim())  # ((1,), 2)

z = x * y  # should throw an exception
print(z.dim())  # the resulting expression had a batch size of 3
```

On the other hand, `x + y` or `dy.cmult(x, y)` does throw an exception, which is the expected behavior.  As far as I have tested, this bug only happens for scalar expressions.

**Comment by chunyang-wen** [comment]
Yes, It is a bug. I'll try to fix.

**Comment by chunyang-wen** [comment]
It seems that `DYNET_ASSERT` is turned off.

```c
    cerr << "left: " << xs[0]->d.bd << " right: " << xs[1]->d.bd << endl; // left: 3 right: 2
    cerr << (xs[1]->d.bd == 1 || xs[1]->d.bd == xs[0]->d.bd) << endl; // 0
    DYNET_ASSERT(xs[1]->d.bd == 1 || xs[1]->d.bd == xs[0]->d.bd, "Failed dimension check in MatrixMultiply::forward");
```

I have tested in `except.h` and `DYNET_ASSERT` is empty.

**Comment by neubig** [comment]
`DYNET_ASSERT` is used for internal programming errors that the user should never accidentally trigger. `DYNET_ARG_CHECK` is used for things the user might trigger, so that `DYNET_ASSERT` should be changed into `DYNET_ARG_CHECK` (and the error should also probably be more explicit).

**Comment by pmichel31415** [comment]
Fixed by #1005
